users/talk.rs

```diff
@@ -0,0 +1,1240 @@
+    ffi::{CStr, CString},
+    fs::{remove_file, File},
+    io::{self, BufRead, Cursor, Error, Read, Write},
+    mem::{size_of, zeroed},
+    net::{
+        self, AddrParseError, Ipv4Addr, SocketAddr, SocketAddrV4, TcpListener, TcpStream, UdpSocket,
+    },
+    process, ptr,
+    sync::{Arc, Mutex},
+    thread,
+    time::{Duration, Instant},
+};
+
+// The size of the buffer for control message fields like l_name, r_name, and r_tty in CtlMsg.
+const BUFFER_SIZE: usize = 12;
+
+// The maximum size for the buffer to store the hostname, based on typical hostname lengths.
+const HOSTNAME_BUFFER_SIZE: usize = 256;
+
+// The maximum number of characters allowed for user input in a single operation.
+const MAX_USER_INPUT_LENGTH: usize = 128;
+const TALK_VERSION: u8 = 1;
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+enum MessageType {
+    LeaveInvite, // leave invitation with server
+    LookUp,      // check for invitation by callee
+    Delete,      // delete invitation by caller
+    Announce,    // announce invitation by caller
+}
+
+impl TryFrom<u8> for MessageType {
+    type Error = ();
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(MessageType::LeaveInvite),
+            1 => Ok(MessageType::LookUp),
+            2 => Ok(MessageType::Delete),
+            3 => Ok(MessageType::Announce),
+            _ => Err(()),
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+enum Answer {
+    Success,          // operation completed properly
+    NotHere,          // callee not logged in
+    Failed,           // operation failed for unexplained reason
+    MachineUnknown,   // caller's machine name unknown
+    PermissionDenied, // callee's tty doesn't permit announce
+    UnknownRequest,   // request has invalid type value
+    BadVersion,       // request has invalid protocol version
+    BadAddr,          // request has invalid addr value
+    BadCtlAddr,       // request has invalid ctl_addr value
+}
+
+impl TryFrom<u8> for Answer {
+    type Error = ();
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(Answer::Success),
+            1 => Ok(Answer::NotHere),
+            2 => Ok(Answer::Failed),
+            3 => Ok(Answer::MachineUnknown),
+            4 => Ok(Answer::PermissionDenied),
+            5 => Ok(Answer::UnknownRequest),
+            6 => Ok(Answer::BadVersion),
+            7 => Ok(Answer::BadAddr),
+            8 => Ok(Answer::BadCtlAddr),
+            _ => Err(()),
+        }
+    }
+}
+
+struct StateLogger {
+    value: String,
+}
+
+impl StateLogger {
+    fn new(initial_value: &str) -> Self {
+        StateLogger {
+            value: initial_value.to_string(),
+        }
+    }
+
+    fn set_state(&mut self, new_value: &str) {
+        if self.value != new_value {
+            println!("{}", new_value);
+            self.value = new_value.to_string();
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum TalkError {
```

Comment:
Are we allowed to use `thiserror`? It would simplify error handling an propagation by quite a bit